### PR TITLE
Add strict mode to scatter_dataset as well as scatter_index

### PR DESCRIPTION
**Pull request by kuenishi**
This PR addresses first half of #7305  by introducing strict mode to existing `scatter_dataset` as well as `scatter_index` to prevent broadcasting underlying dataset. Fix in #7289 is included.

### CODE REVIEW

**Comment by keisukefukuda** [comment]
pfnci, test this please

**Comment by keisukefukuda** [comment]
pfnci, test this please.

File: chainermn/datasets/scatter.py [hunk 1]
```diff
@@ -33,6 +34,8 @@ def scatter_dataset(dataset, comm, root=0, shuffle=False,
             If ``None``, the permutation is changed randomly.
         max_buf_len (int): Max buffer size to be used at broadcasting
             binaries. Must not be larger than 2147483647.
+        strict (bool): Strictly scatters the index and avoid data duplication
```
Comment (keisukefukuda) [inline chainermn/datasets/scatter.py]: I guess we need to document what happens when `len(dataset) % comm.size != 0`. 

The reason why the previous `scatter_dataset()` allows duplication is that all processes must have the same number of samples to keep the iteration count correct.  I'm not 100% sure, but using `strict` option comes with a cost of breaking the restriction ?

Comment (kuenishi) [reply chainermn/datasets/scatter.py]: Updated the document.

**Comment by keisukefukuda** [comment]
pfnci, test this please.

**Comment by iwiwi** [comment]
I don't think `strict` is descriptive enough. How about something like `force_equal_lengths` ?

**Comment by kuenishi** [comment]
Then how about `allow_duplicates` ? I think it's more natural in terms of phrasing software concepts.

**Comment by keisukefukuda** [comment]
+1 for `force_equal_length` or `force_even_split` .

Also, I'm afraid that the docstring is still not enough. I think it should mention *why* and *when* it should be used, and why the default value is `False` (or `True`).

**Comment by kuenishi** [comment]
@keisukefukuda ping

**Comment by keisukefukuda** [comment]
pfnci, test this please.

**Comment by kuenishi** [comment]
jenkins, test this please.

**Comment by keisukefukuda** [comment]
LGTM

**Pull request merged**